### PR TITLE
Move common function attributes check into a function

### PR DESCRIPTION
**Pull request by wilzbach**
From: https://github.com/dlang/dmd/pull/7583/files#r160068084

### CODE REVIEW

File: src/dmd/expression.d [hunk 1]
```diff
@@ -2308,6 +2308,24 @@ extern (C++) abstract class Expression : RootObject
...
+     * and usage of `deprecated` symbols are checked.
+     *
+     * Params:
+     *  sc = scope of the function
+     *  f  = function to be checked
+     * Returns: `true` if error occur.
+     */
+    final bool checkFunctionAttributes(Scope* sc, FuncDeclaration f)
+    {
+        return checkDeprecated(sc, f) ||
```
Comment (JinShil) [inline src/dmd/expression.d]: Note that `checkDeprecated` does not return a `bool` (https://github.com/wilzbach/dmd/blob/226152b34ef76ce8439cc615dc09e5fc9b2d334c/src/dmd/expression.d#L1918).  Perhaps it should, though.  Maybe, as part of this refactoring, add the `bool` return value to make it consistent with other checks.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/expression.d (hunk 1 above)
Comment (JinShil) [reply src/dmd/expression.d]: Or that might be a better change for #7583 to include.  I'll take a look at that.

Comment (JinShil) [reply src/dmd/expression.d]: FYI, I tried to make the change to return `bool` in #7583, but it's too systematic and turns the PR into a monstrosity.  I suggest that after #7583 is merged, we could make an intermediate PR that changes the return type to `bool` and then apply PR on top of that.

Comment (wilzbach) [reply src/dmd/expression.d]: BTW it seems like we should stick with returning `void` for now:

```
Test failed: 
expected:
----
fail_compilation/diag10319.d(25): Error: pure function 'D main' cannot call impure function 'diag10319.foo'
fail_compilation/diag10319.d(25): Error: @safe function 'D main' cannot call @system function 'diag10319.foo'
fail_compilation/diag10319.d(26): Error: pure function 'D main' cannot call impure function 'diag10319.bar!int.bar'
fail_compilation/diag10319.d(26): Error: @safe function 'D main' cannot call @system function 'diag10319.bar!int.bar'
fail_compilation/diag10319.d(25): Error: function `diag10319.foo` is not nothrow
fail_compilation/diag10319.d(26): Error: function `diag10319.bar!int.bar` is not nothrow
fail_compilation/diag10319.d(23): Error: nothrow function `D main` may throw
----
actual:
----
fail_compilation/diag10319.d(25): Error: pure function 'D main' cannot call impure function 'diag10319.foo'
fail_compilation/diag10319.d(26): Error: pure function 'D main' cannot call impure function 'diag10319.bar!int.bar'
fail_compilation/diag10319.d(25): Error: function `diag10319.foo` is not nothrow
fail_compilation/diag10319.d(26): Error: function `diag10319.bar!int.bar` is not nothrow
fail_compilation/diag10319.d(23): Error: nothrow function `D main` may throw
----
```

File: src/dmd/expression.d [hunk 2]
```diff
@@ -2320,6 +2320,28 @@ extern (C++) abstract class Expression : RootObject
...
+     * Checks the attributes of a function.
+     * Purity (`pure`), safety (`@safe`), no GC allocations(`@nogc`)
+     * and usage of `deprecated` and `@disabled`-ed symbols are checked.
+     *
+     * Params:
+     *  sc = scope of the function
+     *  f  = function to be checked
+     * Returns: `true` if error occur.
+     */
+    final bool checkFunctionAttributes(Scope* sc, FuncDeclaration f)
```
Comment (JinShil) [inline src/dmd/expression.d]: I think this would be better as a `private static` function in the `expressionsem.d` module.

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/expression.d (hunk 2 above)
Comment (wilzbach) [reply src/dmd/expression.d]: Done.

**Review by JinShil — APPROVED** [review]

**Review by PetarKirov — APPROVED** [review]

**Comment by ZombineDev** [comment]
Nice, I was thinking of the same refactoring while looking at @JinShil's PR.

**Pull request merged**